macros/contexts/contextNondecimalBase.pl

```diff
@@ -0,0 +1,273 @@
+	#   '&'  => {precedence => .6, associativity => 'left', type => 'bin', string => ' & ',
+	#            class => 'context::Hex::BOP::hex', eval => sub {$_[0] & $_[1]}},
+	#   '|'  => {precedence => .5, associativity => 'left', type => 'bin', string => ' | ',
+	#            class => 'context::Hex::BOP::hex', eval => sub {$_[0] | $_[1]}},
+	#   '^'  => {precedence => .5, associativity => 'left', type => 'bin', string => ' ^ ',
+	#            class => 'context::Hex::BOP::hex', eval => sub {$_[0] ^ $_[1]}},
+	#   '>>' => {precedence => .4, associativity => 'left', type => 'bin', string => ' >> ',
+	#            class => 'context::Hex::BOP::hex', eval => sub {$_[0] >> $_[1]}},
+	#   '<<' => {precedence => .4, associativity => 'left', type => 'bin', string => ' << ',
+	#            class => 'context::Hex::BOP::hex', eval => sub {$_[0] << $_[1]}},
+	#   '~'  => {precedence => 6, associativity => 'left', type => 'unary', string => '~',
+	#            class => 'context::Hex::UOP::not'},
+	# );
+	# $context->{precedence}{Hex} = $context->{precedence}{special};
+	$context->flags->set(limits => [ -1000, 1000, 1 ]);
+	$context->update;
+
+	# main::PG_restricted_eval('sub Hex {context::Hex::Hex->new(@_)}');
+}
+use Data::Dumper;
+
+sub convert {
+	my $value = shift;
+	print Dumper "convert: $value";
+	# Set default options and get passed in options.
+	my %options = (
+		from   => 10,
+		to     => 10,
+		digits => $digits16,
+		@_
+	);
+	my $from   = $options{'from'};
+	my $to     = $options{'to'};
+	my $digits = $options{'digits'};
+	print Dumper "from: $from; to: $to";
+
+	die "The digits option must be an array of characters to use for the digits"
+		unless ref($digits) eq 'ARRAY';
+
+	#
+	# The highest base the digits will support
+	#
+	my $maxBase = scalar(@$digits16);
+
+	die "The base of conversion must be between 2 and $maxBase"
+		unless $to >= 2 && $to <= $maxBase && $from >= 2 && $from <= $maxBase;
+
+	#
+	# Reverse map the digits to base 10 values
+	#
+	my $baseBdigits = { map { ($digits->[$_], $_) } (0 .. $from - 1) };
+
+	#
+	#  Convert to base 10
+	#
+	my $base10;
+	if ($from == 10) {
+		die "The number to convert must consist only of digits: 0,1,2,3,4,5,6,7,8,9"
+			unless $value =~ m/^\d+$/;
+		$base10 = $value;
+	} else {
+		$base10 = 0;
+		foreach my $d (split(//, $value)) {
+			die "The number to convert must consist only of digits: " . join(',', @$digits[ 0 .. $from - 1 ])
+				unless defined($baseBdigits->{$d});
+			$base10 = $base10 * $from + $baseBdigits->{$d};
+		}
+	}
+	return $base10 if $to == 10;
+
+	#
+	#  Convert to desired base
+	#
+	my @base;
+	do {
+		my $d = $base10 % $to;
+		$base10 = ($base10 - $d) / $to;
+		unshift(@base, $digits->[$d]);
+	} while $base10;
+
+	return join('', @base);
+}
+
+# set the base for the context.
+
+use Data::Dumper;
+
+sub setBase {
+	my $b = shift;
+	return Value::Error('The base must be greater than 1 and less than or equal to $max_base')
+		unless $b >= 2 && $b <= scalar(@$digits16);
```

Comment:
There is no `$max_base` variable, and if you are using `$digits16`, you know the highest is 16.

But since your `convert()` allows for specifying a different set of digits, it may be that you want to allow the digits to be specified here.  Then you can tell what the base is and can check that you have the right number of digits (rather than the other way around), and can create the digits array and reverse mapping once instead of doing it in `convert()` each time.  The base, the digits, and the  reverse mapping can be stored as properties in the context itself, rather than as globals.  If you set them up here, you won't need `$digits16` or $digit16` any more (just use `['0' ... '9', 'A' .. 'F']` as the default).

You are right to set the `$context->{pattern}{number}` based on the actual digits (but should only use the ones up to the base requested), though you will need to do `$context->update` after doing so.

If you subclass the Context object to add `convert()` as a method, you should also make `setBase()` a method as well.  Otherwise, you will need to pass the context whose base you are setting as an argument to `setBase()`.

Reply:
> If you subclass the Context object to add convert() as a method, you should also make setBase() a method as well. Otherwise, you will need to pass the context whose base you are setting as an argument to setBase().

I think I'm also missing something here.  This would be a subclass to `Parser::Context` ?